codegen/src/api/metadata.rs

```diff
@@ -0,0 +1,247 @@
+        &self,
+        var: &Variant<PortableForm>,
+        set: &mut HashSet<u32>,
+    ) -> [u8; 32] {
+        let mut bytes = vec![MetadataHashableIDs::Variant as u8];
+
+        bytes.extend(var.name().as_bytes());
+        for field in var.fields() {
+            bytes.extend(self.get_field_uid(field, set));
+        }
+
+        MetadataHashable::hash(&bytes)
+    }
+
+    fn get_type_def_uid(
+        &self,
+        ty_def: &TypeDef<PortableForm>,
+        set: &mut HashSet<u32>,
+    ) -> [u8; 32] {
+        let mut bytes = vec![MetadataHashableIDs::TypeDef as u8];
+
+        let data = match ty_def {
+            TypeDef::Composite(composite) => {
+                let mut bytes = Vec::new();
+                for field in composite.fields() {
+                    bytes.extend(self.get_field_uid(field, set));
+                }
+                bytes
+            }
+            TypeDef::Variant(variant) => {
+                let mut bytes = Vec::new();
+                for var in variant.variants() {
+                    bytes.extend(self.get_variant_uid(var, set));
+                }
+                bytes
+            }
+            TypeDef::Sequence(sequence) => {
+                let mut bytes = Vec::new();
+                bytes.extend(self.get_type_uid_internal(sequence.type_param().id(), set));
+                bytes
+            }
+            TypeDef::Array(array) => {
+                let mut bytes = Vec::new();
+                bytes.extend(array.len().to_be_bytes());
+                bytes.extend(self.get_type_uid_internal(array.type_param().id(), set));
+                bytes
+            }
+            TypeDef::Tuple(tuple) => {
+                let mut bytes = Vec::new();
+                for field in tuple.fields() {
+                    bytes.extend(self.get_type_uid_internal(field.id(), set));
+                }
+                bytes
+            }
+            TypeDef::Primitive(primitive) => {
+                let mut bytes = Vec::new();
+                bytes.extend(primitive.encode());
+                bytes
+            }
+            TypeDef::Compact(compact) => {
+                let mut bytes = Vec::new();
+                bytes.extend(self.get_type_uid_internal(compact.type_param().id(), set));
+                bytes
+            }
+            TypeDef::BitSequence(bitseq) => {
+                let mut bytes = Vec::new();
+                bytes.extend(
+                    self.get_type_uid_internal(bitseq.bit_order_type().id(), set),
+                );
+                bytes.extend(
+                    self.get_type_uid_internal(bitseq.bit_store_type().id(), set),
+                );
+                bytes
+            }
+        };
+        bytes.extend(data);
+        MetadataHashable::hash(&bytes)
+    }
+
+    fn get_type_uid_internal(&self, id: u32, set: &mut HashSet<u32>) -> [u8; 32] {
+        lazy_static! {
+            static ref CACHED_UID: Mutex<HashMap<u32, [u8; 32]>> =
```

Comment:
I don't think we'd want this to exist for the lifetime of the application. It should probably be fresh and empty each time we hash some new pallet or something, shouldn't it? Otherwise we'll hit "recursive" IDs for the first time, and stop recursing rather than going at least some of the way into them first I think?